File: src/peripherals/common/memory.js
// @ts-check
/** Base memory class — flat byte array with an absolute base address.
 * Provides little-endian read/write at 8/16/32-bit widths. */
export class Memory {
  /** @param {Uint8Array} data @param {number} base */
  constructor(data, base) {
    ((this.data = data), (this.base = base));
  }
  /** @returns {number} absolute base address */
  get baseAddr() {
    return this.base;
  }
  /** @param {number} addr @returns {boolean} */
  contains(addr) {
    return addr >= this.base && addr < this.base + this.data.length;
  }
  /** @param {number} addr @returns {number} */
  readUint8(addr) {
    let { data: tmpVal, base: idxVal } = this;
    return tmpVal[addr - idxVal];
  }
  /** @param {number} addr @returns {number} */
  readUint16(addr) {
    let { data: tmpVal, base: idxVal } = this;
    return tmpVal[addr - idxVal] | (tmpVal[addr - idxVal + 1] << 8);
  }
  /** @param {number} addr @returns {number} */
  readUint32(addr) {
    let { data: tmpVal, base: idxVal } = this,
      off = addr - idxVal;
    return (
      (tmpVal[off] |
        (tmpVal[off + 1] << 8) |
        (tmpVal[off + 2] << 16) |
        (tmpVal[off + 3] << 24)) >>>
      0
    );
  }
  /** @param {number} addr @param {number} val */
  writeUint8(addr, val) {
    let { data: idxVal, base: ClockEvent } = this;
    idxVal[addr - ClockEvent] = val;
  }
  /** @param {number} addr @param {number} val */
  writeUint16(addr, val) {
    let { data: idxVal, base: ClockEvent } = this;
    ((idxVal[addr - ClockEvent] = 255 & val),
      (idxVal[addr - ClockEvent + 1] = (val >> 8) & 255));
  }
  /** @param {number} addr @param {number} val */
  writeUint32(addr, val) {
    let { data: idxVal, base: ClockEvent } = this,
      off = addr - ClockEvent;
    ((idxVal[off] = 255 & val),
      (idxVal[off + 1] = (val >> 8) & 255),
      (idxVal[off + 2] = (val >> 16) & 255),
      (idxVal[off + 3] = (val >> 24) & 255));
  }
  /** @param {Uint8Array} src @param {number} srcBase */
  set(src, srcBase) {
    this.data.set(src, srcBase - this.base);
  }
  /** @param {Memory} src @param {number} srcAddr @param {number} dstAddr @param {number} len */
  copy(src, srcAddr, dstAddr, len) {
    this.data.set(
      src.data.subarray(
        srcAddr - src.base,
        srcAddr - src.base + len,
      ),
      dstAddr - this.base,
    );
  }
  /** @param {number} base @param {number} len @returns {Memory} */
  createView(base, len) {
    return new Memory(this.data.subarray(len, len + base), base);
  }
  /** @param {number} base @returns {Memory} */
  remap(base) {
    return new Memory(this.data, base);
  }
}

/**
 * Read-only memory region. Writes are blocked unless
 * `ReadonlyMemory.override` is true (used by GDB flash writes).
 */
export class ReadonlyMemory extends Memory {
  /** @param {number} addr @param {number} val */
  writeUint8(addr, val) {
    if (ReadonlyMemory.override) return super.writeUint8(addr, val);
    console.error("Invalid write to read-only memory at", addr.toString(16));
  }
  /** @param {number} addr @param {number} val */
  writeUint16(addr, val) {
    if (ReadonlyMemory.override) return super.writeUint16(addr, val);
    console.error("Invalid write to read-only memory at", addr.toString(16));
  }
  /** @param {number} addr @param {number} val */
  writeUint32(addr, val) {
    if (ReadonlyMemory.override) return super.writeUint32(addr, val);
    console.error("Invalid write to read-only memory at", addr.toString(16));
  }
  /** @param {number} base @param {number} len @returns {ReadonlyMemory} */
  createView(base, len) {
    return new ReadonlyMemory(
      this.data.subarray(len, len + base),
      base,
    );
  }
}
/** @type {boolean} Global override flag — set by GDB session during flash writes */
ReadonlyMemory.override = false;

/** Stub memory: reads return a fixed default, writes are dropped. */
export class InvalidMemory {
  /** @param {number} [defaultValue=0xFFFFFFFF] */
  constructor(defaultValue = 0xffffffff) {
    this.defaultValue = defaultValue;
  }
  /** @returns {number} */
  readUint8() { return 255 & this.defaultValue; }
  /** @returns {number} */
  readUint16() { return 65535 & this.defaultValue; }
  /** @returns {number} */
  readUint32() { return 0 | this.defaultValue; }
  writeUint8() {}
  writeUint16() {}
  writeUint32() {}
}

/** Address-shifting wrapper: translates addresses by a fixed delta before forwarding. */
export class MemoryTranslator {
  /** @param {Memory} base @param {number} delta */
  constructor(base, delta) {
    ((this.base = base), (this.delta = delta));
  }
  /** @param {number} addr @returns {number} */
  readUint8(addr) { return this.base.readUint8(addr - this.delta); }
  /** @param {number} addr @returns {number} */
  readUint16(addr) { return this.base.readUint16(addr - this.delta); }
  /** @param {number} addr @returns {number} */
  readUint32(addr) { return this.base.readUint32(addr - this.delta); }
  /** @param {number} addr @param {number} val */
  writeUint8(addr, val) { this.base.writeUint8(addr - this.delta, val); }
  /** @param {number} addr @param {number} val */
  writeUint16(addr, val) { this.base.writeUint16(addr - this.delta, val); }
  /** @param {number} addr @param {number} val */
  writeUint32(addr, val) { this.base.writeUint32(addr - this.delta, val); }
}

/** Byte-reversed addressing (for SRAM1 reverse window). */
export class ReverseMemory {
  /** @param {Uint8Array} data @param {number} base */
  constructor(data, base) {
    ((this.data = data), (this.base = base));
  }
  /** @param {number} addr @returns {number} byte-reversed offset */
  translateAddress(addr) {
    let off = addr - this.base;
    return this.data.length - 4 - (0xfffffffc & off) + (3 & off);
  }
  /** @param {number} addr @returns {number} */
  readUint8(addr) { return this.data[this.translateAddress(addr)]; }
  /** @param {number} addr @returns {number} */
  readUint16(addr) {
    let lo = this.translateAddress(addr),
      hi = this.translateAddress(addr + 1);
    return this.data[lo] | (this.data[hi] << 8);
  }
  /** @param {number} addr @returns {number} */
  readUint32(addr) {
    let off = this.translateAddress(addr);
    return (
      (this.data[off] |
        (this.data[off + 1] << 8) |
        (this.data[off + 2] << 16) |
        (this.data[off + 3] << 24)) >>> 0
    );
  }
  /** @param {number} addr @param {number} val */
  writeUint8(addr, val) { this.data[this.translateAddress(addr)] = val; }
  /** @param {number} addr @param {number} val */
  writeUint16(addr, val) {
    let lo = this.translateAddress(addr),
      hi = this.translateAddress(addr + 1);
    ((this.data[lo] = 255 & val),
      (this.data[hi] = (val >> 8) & 255));
  }
  /** @param {number} addr @param {number} val */
  writeUint32(addr, val) {
    let off = this.translateAddress(addr);
    ((this.data[off] = 255 & val),
      (this.data[off + 1] = (val >> 8) & 255),
      (this.data[off + 2] = (val >> 16) & 255),
      (this.data[off + 3] = (val >> 24) & 255));
  }
}

/** MMU-table-driven address translation (used for PSRAM region). */
export class MMUMemory extends Memory {
  /** @param {Uint8Array} data @param {number} base */
  constructor(data, base) {
    (super(data, base),
      (this.table = new Uint32Array(0)),
      (this.tableOffset = 0));
  }
  /** @type {Uint32Array} */  table;
  /** @type {number} */  tableOffset;
  /** @param {Uint32Array} table @param {number} offset */
  setTable(table, offset) {
    this.table = table;
    this.tableOffset = offset;
  }
  /** @param {number} addr @returns {number} physical address */
  mapAddress(addr) {
    let off = addr - this.baseAddr,
      idx = off >>> 15,
      entry = this.table[this.tableOffset + idx] ?? idx;
    return this.baseAddr + (entry << 15) + (off & 32767);
  }
  /** @param {number} addr @returns {number} */
  readUint8(addr) { return ((addr = this.mapAddress(addr)), super.readUint8(addr)); }
  /** @param {number} addr @returns {number} */
  readUint16(addr) { return ((addr = this.mapAddress(addr)), super.readUint16(addr)); }
  /** @param {number} addr @returns {number} */
  readUint32(addr) { return ((addr = this.mapAddress(addr)), super.readUint32(addr)); }
  /** @param {number} addr @param {number} val */
  writeUint8(addr, val) { ((addr = this.mapAddress(addr)), super.writeUint8(addr, val)); }
  /** @param {number} addr @param {number} val */
  writeUint16(addr, val) { ((addr = this.mapAddress(addr)), super.writeUint16(addr, val)); }
  /** @param {number} addr @param {number} val */
  writeUint32(addr, val) { ((addr = this.mapAddress(addr)), super.writeUint32(addr, val)); }
}

// ─── Page Table & MMIO Handler Registry ───────────────────────────────────

export const PAGE_SHIFT = 12;
export const PAGE_SIZE = 1 << PAGE_SHIFT;       // 4096
export const PAGE_ENTRIES = 0x100000;            // 1,048,576 (covers 4GB)
export const PT_SIZE = PAGE_ENTRIES * 2;         // 2,097,152 u32 entries

export const PTE_TYPE_INVALID = 0;
export const PTE_TYPE_RAM = 1;
export const PTE_TYPE_MMIO = 2;
export const PTE_TYPE_FLASH = 3;

/** 4GB page table — 1M entries × 2 u32 each (type + data). */
export class PageTable {
  constructor() {
    this.table = new Uint32Array(PT_SIZE);
  }
  /** @param {number} addr @returns {number} */
  pageIndex(addr) { return addr >>> PAGE_SHIFT; }
  /** @param {number} addr @returns {number} PTE type */
  getType(addr) { return this.table[this.pageIndex(addr) * 2]; }
  /** @param {number} addr @returns {number} PTE data (RAM offset or MMIO handler ID) */
  getData(addr) { return this.table[this.pageIndex(addr) * 2 + 1]; }
  /** @param {number} baseAddr @param {number} size @param {number} type @param {number} data */
  setRange(baseAddr, size, type, data) {
    const start = this.pageIndex(baseAddr);
    const end = this.pageIndex(baseAddr + size - 1);
    for (let i = start; i <= end; i++) {
      this.table[i * 2] = type;
      this.table[i * 2 + 1] = data;
    }
  }
  /** @param {number} addr @param {number} type @param {number} data */
  setPage(addr, type, data) {
    const i = this.pageIndex(addr);
    this.table[i * 2] = type;
    this.table[i * 2 + 1] = data;
  }
}

/** Registry of MMIO handler objects, indexed by handler ID. */
export class MMIOHandlerRegistry {
  constructor() { this.handlers = []; }
  /** @param {object} handler @returns {number} handler ID */
  register(handler) {
    const id = this.handlers.length;
    this.handlers.push(handler);
    return id;
  }
  /** @param {number} id @returns {object|undefined} */
  get(id) { return this.handlers[id]; }
  /** @param {number} handlerId @param {number} addr @param {number} size @returns {number} */
  read(handlerId, addr, size) {
    const h = this.handlers[handlerId];
    if (!h) return 0;
    switch (size) {
      case 1: return h.mmio_read?.(addr, 1) ?? h.readUint8?.(addr) ?? 0;
      case 2: return h.mmio_read?.(addr, 2) ?? h.readUint16?.(addr) ?? 0;
      case 4: return h.mmio_read?.(addr, 4) ?? h.readUint32?.(addr) ?? 0;
      default: return 0;
    }
  }
  /** @param {number} handlerId @param {number} addr @param {number} val @param {number} size */
  write(handlerId, addr, val, size) {
    const h = this.handlers[handlerId];
    if (!h) return;
    switch (size) {
      case 1: h.mmio_write?.(addr, val, 1) ?? h.writeUint8?.(addr, val); break;
      case 2: h.mmio_write?.(addr, val, 2) ?? h.writeUint16?.(addr, val); break;
      case 4: h.mmio_write?.(addr, val, 4) ?? h.writeUint32?.(addr, val); break;
    }
  }
}

File: engine-wasm/src/peripherals/common/memory.rs
use crate::peripherals::types::*;
use core::sync::atomic::{AtomicBool, Ordering};

pub const PAGE_SHIFT: u32 = 12;
pub const PAGE_SIZE: u32 = 1 << PAGE_SHIFT;
pub const PAGE_ENTRIES: u32 = 0x100000;
pub const PT_SIZE: u32 = PAGE_ENTRIES * 2;

pub const PTE_TYPE_INVALID: u32 = 0;
pub const PTE_TYPE_RAM: u32 = 1;
pub const PTE_TYPE_MMIO: u32 = 2;

pub static READONLY_OVERRIDE: AtomicBool = AtomicBool::new(false);

pub struct Memory {
    pub data: *mut u8,
    pub len: u32,
    pub base: u32,
}

impl Memory {
    pub fn new(data: *mut u8, len: u32, base: u32) -> Self {
        Memory { data, len, base }
    }

    pub fn base_addr(&self) -> u32 {
        self.base
    }

    pub fn contains(&self, addr: u32) -> bool {
        addr >= self.base && addr < self.base + self.len
    }

    pub fn read_u8(&self, addr: u32) -> u32 {
        let offset = (addr - self.base) as usize;
        unsafe { *self.data.add(offset) as u32 }
    }

    pub fn read_u16(&self, addr: u32) -> u32 {
        let offset = (addr - self.base) as usize;
        unsafe {
            let low = *self.data.add(offset) as u32;
            let high = (*self.data.add(offset + 1) as u32) << 8;
            low | high
        }
    }

    pub fn read_u32(&self, addr: u32) -> u32 {
        let offset = (addr - self.base) as usize;
        unsafe {
            let b0 = *self.data.add(offset) as u32;
            let b1 = (*self.data.add(offset + 1) as u32) << 8;
            let b2 = (*self.data.add(offset + 2) as u32) << 16;
            let b3 = (*self.data.add(offset + 3) as u32) << 24;
            b0 | b1 | b2 | b3
        }
    }

    pub fn write_u8(&mut self, addr: u32, val: u32) {
        let offset = (addr - self.base) as usize;
        unsafe { *self.data.add(offset) = (val & 0xFF) as u8 }
    }

    pub fn write_u16(&mut self, addr: u32, val: u32) {
        let offset = (addr - self.base) as usize;
        unsafe {
            *self.data.add(offset) = (val & 0xFF) as u8;
            *self.data.add(offset + 1) = ((val >> 8) & 0xFF) as u8;
        }
    }

    pub fn write_u32(&mut self, addr: u32, val: u32) {
        let offset = (addr - self.base) as usize;
        unsafe {
            *self.data.add(offset) = (val & 0xFF) as u8;
            *self.data.add(offset + 1) = ((val >> 8) & 0xFF) as u8;
            *self.data.add(offset + 2) = ((val >> 16) & 0xFF) as u8;
            *self.data.add(offset + 3) = ((val >> 24) & 0xFF) as u8;
        }
    }

    pub fn set(&mut self, src: &[u8], dest_addr: u32) {
        let offset = (dest_addr - self.base) as usize;
        unsafe {
            core::ptr::copy_nonoverlapping(src.as_ptr(), self.data.add(offset), src.len());
        }
    }

    pub fn copy(&mut self, src: &Memory, src_addr: u32, dest_addr: u32, len: u32) {
        let src_offset = (src_addr - src.base) as usize;
        let dest_offset = (dest_addr - self.base) as usize;
        unsafe {
            core::ptr::copy_nonoverlapping(src.data.add(src_offset), self.data.add(dest_offset), len as usize);
        }
    }

    pub fn create_view(&self, base: u32, offset: u32, len: u32) -> Memory {
        let ptr = unsafe { self.data.add(offset as usize) };
        Memory::new(ptr, len, base)
    }

    pub fn remap(&self, base: u32) -> Memory {
        Memory::new(self.data, self.len, base)
    }
}

pub struct ReadonlyMemory {
    pub mem: Memory,
}

impl ReadonlyMemory {
    pub fn new(data: *mut u8, len: u32, base: u32) -> Self {
        ReadonlyMemory { mem: Memory::new(data, len, base) }
    }

    pub fn base_addr(&self) -> u32 {
        self.mem.base_addr()
    }

    pub fn contains(&self, addr: u32) -> bool {
        self.mem.contains(addr)
    }

    pub fn read_u8(&self, addr: u32) -> u32 {
        self.mem.read_u8(addr)
    }

    pub fn read_u16(&self, addr: u32) -> u32 {
        self.mem.read_u16(addr)
    }

    pub fn read_u32(&self, addr: u32) -> u32 {
        self.mem.read_u32(addr)
    }

    pub fn write_u8(&mut self, addr: u32, val: u32) {
        if READONLY_OVERRIDE.load(Ordering::Relaxed) {
            return self.mem.write_u8(addr, val);
        }
    }

    pub fn write_u16(&mut self, addr: u32, val: u32) {
        if READONLY_OVERRIDE.load(Ordering::Relaxed) {
            return self.mem.write_u16(addr, val);
        }
    }

    pub fn write_u32(&mut self, addr: u32, val: u32) {
        if READONLY_OVERRIDE.load(Ordering::Relaxed) {
            return self.mem.write_u32(addr, val);
        }
    }

    pub fn set(&mut self, src: &[u8], dest_addr: u32) {
        self.mem.set(src, dest_addr)
    }

    pub fn copy(&mut self, src: &Memory, src_addr: u32, dest_addr: u32, len: u32) {
        self.mem.copy(src, src_addr, dest_addr, len)
    }

    pub fn create_view(&self, base: u32, offset: u32, len: u32) -> ReadonlyMemory {
        let ptr = unsafe { self.mem.data.add(offset as usize) };
        ReadonlyMemory::new(ptr, len, base)
    }

    pub fn remap(&self, base: u32) -> Memory {
        self.mem.remap(base)
    }
}

pub struct InvalidMemory {
    pub default_value: u32,
}

impl InvalidMemory {
    pub fn new(default_value: u32) -> Self {
        InvalidMemory { default_value }
    }

    pub fn read_u8(&self) -> u32 {
        0xFF & self.default_value
    }

    pub fn read_u16(&self) -> u32 {
        0xFFFF & self.default_value
    }

    pub fn read_u32(&self) -> u32 {
        0 | self.default_value
    }

    pub fn write_u8(&mut self) {}

    pub fn write_u16(&mut self) {}

    pub fn write_u32(&mut self) {}
}

pub struct MemoryTranslator {
    pub base: *mut Memory,
    pub delta: u32,
}

impl MemoryTranslator {
    pub fn new(base: *mut Memory, delta: u32) -> Self {
        MemoryTranslator { base, delta }
    }

    pub fn read_u8(&self, addr: u32) -> u32 {
        unsafe { (*self.base).read_u8(addr - self.delta) }
    }

    pub fn read_u16(&self, addr: u32) -> u32 {
        unsafe { (*self.base).read_u16(addr - self.delta) }
    }

    pub fn read_u32(&self, addr: u32) -> u32 {
        unsafe { (*self.base).read_u32(addr - self.delta) }
    }

    pub fn write_u8(&mut self, addr: u32, val: u32) {
        unsafe { (*self.base).write_u8(addr - self.delta, val) }
    }

    pub fn write_u16(&mut self, addr: u32, val: u32) {
        unsafe { (*self.base).write_u16(addr - self.delta, val) }
    }

    pub fn write_u32(&mut self, addr: u32, val: u32) {
        unsafe { (*self.base).write_u32(addr - self.delta, val) }
    }
}

pub struct ReverseMemory {
    pub data: *mut u8,
    pub len: u32,
    pub base: u32,
}

impl ReverseMemory {
    pub fn new(data: *mut u8, len: u32, base: u32) -> Self {
        ReverseMemory { data, len, base }
    }

    pub fn translate_address(&self, addr: u32) -> u32 {
        let offset = addr - self.base;
        self.len - 4 - (0xFFFFFFFC & offset) + (3 & offset)
    }

    pub fn read_u8(&self, addr: u32) -> u32 {
        let idx = self.translate_address(addr) as usize;
        unsafe { *self.data.add(idx) as u32 }
    }

    pub fn read_u16(&self, addr: u32) -> u32 {
        let idx0 = self.translate_address(addr) as usize;
        let idx1 = self.translate_address(addr + 1) as usize;
        unsafe {
            let low = *self.data.add(idx0) as u32;
            let high = (*self.data.add(idx1) as u32) << 8;
            low | high
        }
    }

    pub fn read_u32(&self, addr: u32) -> u32 {
        let base = self.translate_address(addr) as usize;
        unsafe {
            let b0 = *self.data.add(base) as u32;
            let b1 = (*self.data.add(base + 1) as u32) << 8;
            let b2 = (*self.data.add(base + 2) as u32) << 16;
            let b3 = (*self.data.add(base + 3) as u32) << 24;
            b0 | b1 | b2 | b3
        }
    }

    pub fn write_u8(&mut self, addr: u32, val: u32) {
        let idx = self.translate_address(addr) as usize;
        unsafe { *self.data.add(idx) = val as u8 }
    }

    pub fn write_u16(&mut self, addr: u32, val: u32) {
        let idx0 = self.translate_address(addr) as usize;
        let idx1 = self.translate_address(addr + 1) as usize;
        unsafe {
            *self.data.add(idx0) = (val & 0xFF) as u8;
            *self.data.add(idx1) = ((val >> 8) & 0xFF) as u8;
        }
    }

    pub fn write_u32(&mut self, addr: u32, val: u32) {
        let base = self.translate_address(addr) as usize;
        unsafe {
            *self.data.add(base) = (val & 0xFF) as u8;
            *self.data.add(base + 1) = ((val >> 8) & 0xFF) as u8;
            *self.data.add(base + 2) = ((val >> 16) & 0xFF) as u8;
            *self.data.add(base + 3) = ((val >> 24) & 0xFF) as u8;
        }
    }
}

pub struct MMUMemory {
    pub mem: Memory,
    pub table: *mut u32,
    pub table_len: u32,
    pub table_offset: u32,
}

impl MMUMemory {
    pub fn new(data: *mut u8, len: u32, base: u32) -> Self {
        MMUMemory {
            mem: Memory::new(data, len, base),
            table: core::ptr::null_mut(),
            table_len: 0,
            table_offset: 0,
        }
    }

    pub fn base_addr(&self) -> u32 {
        self.mem.base_addr()
    }

    pub fn contains(&self, addr: u32) -> bool {
        self.mem.contains(addr)
    }

    pub fn set_table(&mut self, table: *mut u32, table_len: u32, offset: u32) {
        self.table = table;
        self.table_len = table_len;
        self.table_offset = offset;
    }

    pub fn map_address(&self, addr: u32) -> u32 {
        let offset = addr - self.base_addr();
        let idx = offset >> 15;
        let entry = if (self.table_offset + idx) < self.table_len {
            unsafe { *self.table.add((self.table_offset + idx) as usize) }
        } else {
            idx
        };
        self.base_addr() + (entry << 15) + (offset & 32767)
    }

    pub fn read_u8(&self, addr: u32) -> u32 {
        let mapped = self.map_address(addr);
        self.mem.read_u8(mapped)
    }

    pub fn read_u16(&self, addr: u32) -> u32 {
        let mapped = self.map_address(addr);
        self.mem.read_u16(mapped)
    }

    pub fn read_u32(&self, addr: u32) -> u32 {
        let mapped = self.map_address(addr);
        self.mem.read_u32(mapped)
    }

    pub fn write_u8(&mut self, addr: u32, val: u32) {
        let mapped = self.map_address(addr);
        self.mem.write_u8(mapped, val)
    }

    pub fn write_u16(&mut self, addr: u32, val: u32) {
        let mapped = self.map_address(addr);
        self.mem.write_u16(mapped, val)
    }

    pub fn write_u32(&mut self, addr: u32, val: u32) {
        let mapped = self.map_address(addr);
        self.mem.write_u32(mapped, val)
    }

    pub fn set(&mut self, src: &[u8], dest_addr: u32) {
        self.mem.set(src, dest_addr)
    }

    pub fn copy(&mut self, src: &Memory, src_addr: u32, dest_addr: u32, len: u32) {
        self.mem.copy(src, src_addr, dest_addr, len)
    }

    pub fn create_view(&self, base: u32, offset: u32, len: u32) -> Memory {
        self.mem.create_view(base, offset, len)
    }

    pub fn remap(&self, base: u32) -> Memory {
        self.mem.remap(base)
    }
}

pub struct PageTable {
    pub table: *mut u32,
    pub len: u32,
}

impl PageTable {
    pub fn new(table: *mut u32, len: u32) -> Self {
        PageTable { table, len }
    }

    pub fn page_index(&self, addr: u32) -> u32 {
        addr >> PAGE_SHIFT
    }

    pub fn get_type(&self, addr: u32) -> u32 {
        let i = self.page_index(addr) * 2;
        unsafe { *self.table.add(i as usize) }
    }

    pub fn get_data(&self, addr: u32) -> u32 {
        let i = self.page_index(addr) * 2 + 1;
        unsafe { *self.table.add(i as usize) }
    }

    pub fn set_range(&mut self, base_addr: u32, size: u32, ptype: u32, data: u32) {
        let start = self.page_index(base_addr);
        let end = self.page_index(base_addr + size - 1);
        for i in start..=end {
            unsafe {
                *self.table.add((i * 2) as usize) = ptype;
                *self.table.add((i * 2 + 1) as usize) = data;
            }
        }
    }

    pub fn set_page(&mut self, addr: u32, ptype: u32, data: u32) {
        let i = self.page_index(addr) * 2;
        unsafe {
            *self.table.add(i as usize) = ptype;
            *self.table.add((i + 1) as usize) = data;
        }
    }
}

pub type MmioReadFn = unsafe fn(*mut u8, u32, u32) -> u32;
pub type MmioWriteFn = unsafe fn(*mut u8, u32, u32, u32);
pub type ReadFn = unsafe fn(*mut u8, u32) -> u32;
pub type WriteFn = unsafe fn(*mut u8, u32, u32);

#[derive(Clone, Copy)]
pub struct MMIOHandler {
    pub context: *mut u8,
    pub mmio_read: Option<MmioReadFn>,
    pub mmio_write: Option<MmioWriteFn>,
    pub read_u8: Option<ReadFn>,
    pub read_u16: Option<ReadFn>,
    pub read_u32: Option<ReadFn>,
    pub write_u8: Option<WriteFn>,
    pub write_u16: Option<WriteFn>,
    pub write_u32: Option<WriteFn>,
}

impl MMIOHandler {
    pub const fn empty() -> Self {
        MMIOHandler {
            context: core::ptr::null_mut(),
            mmio_read: None,
            mmio_write: None,
            read_u8: None,
            read_u16: None,
            read_u32: None,
            write_u8: None,
            write_u16: None,
            write_u32: None,
        }
    }
}

pub const MAX_HANDLERS: usize = 256;

pub struct MMIOHandlerRegistry {
    pub handlers: [MMIOHandler; MAX_HANDLERS],
    pub count: u32,
}

impl MMIOHandlerRegistry {
    pub fn new() -> Self {
        MMIOHandlerRegistry {
            handlers: [MMIOHandler::empty(); MAX_HANDLERS],
            count: 0,
        }
    }

    pub fn register(&mut self, handler: MMIOHandler) -> u32 {
        let id = self.count;
        self.handlers[id as usize] = handler;
        self.count += 1;
        id
    }

    pub fn get(&self, id: u32) -> Option<&MMIOHandler> {
        if id < self.count {
            Some(&self.handlers[id as usize])
        } else {
            None
        }
    }

    pub fn read(&self, handler_id: u32, addr: u32, size: u32) -> u32 {
        if handler_id >= self.count {
            return 0;
        }
        let h = &self.handlers[handler_id as usize];
        match size {
            1 => {
                if let Some(f) = h.mmio_read {
                    unsafe { f(h.context, addr, 1) }
                } else if let Some(f) = h.read_u8 {
                    unsafe { f(h.context, addr) }
                } else {
                    0
                }
            }
            2 => {
                if let Some(f) = h.mmio_read {
                    unsafe { f(h.context, addr, 2) }
                } else if let Some(f) = h.read_u16 {
                    unsafe { f(h.context, addr) }
                } else {
                    0
                }
            }
            4 => {
                if let Some(f) = h.mmio_read {
                    unsafe { f(h.context, addr, 4) }
                } else if let Some(f) = h.read_u32 {
                    unsafe { f(h.context, addr) }
                } else {
                    0
                }
            }
            _ => 0,
        }
    }

    pub fn write(&mut self, handler_id: u32, addr: u32, val: u32, size: u32) {
        if handler_id >= self.count {
            return;
        }
        let h = &self.handlers[handler_id as usize];
        match size {
            1 => {
                if let Some(f) = h.mmio_write {
                    unsafe { f(h.context, addr, val, 1) }
                } else if let Some(f) = h.write_u8 {
                    unsafe { f(h.context, addr, val) }
                }
            }
            2 => {
                if let Some(f) = h.mmio_write {
                    unsafe { f(h.context, addr, val, 2) }
                } else if let Some(f) = h.write_u16 {
                    unsafe { f(h.context, addr, val) }
                }
            }
            4 => {
                if let Some(f) = h.mmio_write {
                    unsafe { f(h.context, addr, val, 4) }
                } else if let Some(f) = h.write_u32 {
                    unsafe { f(h.context, addr, val) }
                }
            }
            _ => {}
        }
    }
}
